src/storage/cache/queue.rs

```diff
@@ -0,0 +1,203 @@
+use core::marker::PhantomData;
+use std::{
+    ptr::{null_mut, NonNull},
+    sync::atomic::{AtomicPtr, Ordering::*},
+};
+
+pub struct Queue<T> {
+    size: usize,
+    head: AtomicPtr<Node<T>>,
+    tail: AtomicPtr<Node<T>>,
+}
+
+impl<T> Queue<T> {
+    pub fn new() -> Self {
+        let node = Node::new(None);
+        let node = Box::into_raw(Box::new(node));
+        Self {
+            head: AtomicPtr::new(node),
+            tail: AtomicPtr::new(node),
+            size: 0,
+        }
+    }
+
+    #[inline(always)]
+    pub fn push(&mut self, item: T) {
+        let node = Node::new(Some(item));
+        let node_ptr = Box::into_raw(Box::new(node));
+        let prev_tail = self.tail.load(Relaxed);
+        unsafe {
```

Comment:
One of the best practices in Rust is to add a comment explaining why `unsafe` blocks are safe in the particular context.

Reply:
Noted, thanks @rushmorem, I have resorted to using the LinkedList from the std lib, so removing the usage of unsafe here.